src/java/voldemort/store/rocksdb/RocksDbStorageEngine.java

```diff
@@ -0,0 +1,490 @@
+             */
+            List<Versioned<byte[]>> currentValues;
+            try {
+                byte[] result = getRocksDB().get(key.get());
+                if(result != null) {
+                    currentValues = StoreBinaryFormat.fromByteArray(result);
+                } else {
+                    currentValues = Collections.emptyList();
+                }
+            } catch(RocksDBException e) {
+                logger.error(e);
+                throw new PersistenceFailureException(e);
+            }
+            if(currentValues.size() > 0) {
+                // compare vector clocks and throw out old ones, for updates
+                Iterator<Versioned<byte[]>> iter = currentValues.iterator();
+                while(iter.hasNext()) {
+                    Versioned<byte[]> curr = iter.next();
+                    Occurred occured = value.getVersion().compare(curr.getVersion());
+                    if(occured == Occurred.BEFORE) {
+                        throw new ObsoleteVersionException("Key "
+                                                           + new String(hexCodec.encode(key.get()))
+                                                           + " "
+                                                           + value.getVersion().toString()
+                                                           + " is obsolete, it is no greater than the current version of "
+                                                           + curr.getVersion().toString() + ".");
+                    } else if(occured == Occurred.AFTER) {
+                        iter.remove();
+                    }
+                }
+            } else {
+                // if value does not exist add the value from put request to
+                // existing values
+                currentValues = new ArrayList<Versioned<byte[]>>(1);
+            }
+            currentValues.add(value);
+
+            try {
+                getRocksDB().put(key.get(), StoreBinaryFormat.toByteArray(currentValues));
+            } catch(RocksDBException e) {
+                logger.error(e);
+                throw new PersistenceFailureException(e);
+            } finally {
+                if(logger.isTraceEnabled()) {
+                    logger.trace("Completed PUT (" + getName() + ") to key " + key + " (keyRef: "
+                                 + System.identityHashCode(key) + " value " + value + " in "
+                                 + (System.nanoTime() - startTimeNs) + " ns at "
+                                 + System.currentTimeMillis());
+                }
+            }
+        }
+    }
+
+    @Override
+    public boolean delete(ByteArray key, Version version) throws PersistenceFailureException {
+
+        StoreUtils.assertValidKey(key);
+
+        long startTimeNs = -1;
+
+        if(logger.isTraceEnabled())
+            startTimeNs = System.nanoTime();
+
+        synchronized(this.locks.lockFor(key.get())) {
```

Comment:
so for put and delete, we do RMW at our level.. is there a way to push this down to rocksdb (merge operator?)